_posts/2020-07-29-springboot-logging.md

```diff
@@ -0,0 +1,336 @@
+```
+Alternatively, we can specify our package in the application.properties :
+
+```properties
+logging.level.org.springframework=ERROR 
+logging.level.io.app=TRACE
+```
+
+### Log To File For Archival Or Shipping To Log Aggregators
+We can write our logs to a file path by setting only one of the properties logging.file.name or logging.file.path in our application.properties. By default, for file output, the log level is set to info. 
+
+```
+# Output to a file named application.log. 
+logging.file.name=application.log
+```
+```
+# Output to a file named spring.log in path /Users
+logging.file.path=/Users
+```
+If both properties are set, only the logging.file.name takes effect. 
+
+Note: The name of these properties has changed in spring 2.2 onwards but the official documentation does not yet reflect this. Our example is working with version 2.3.2.RELEASE. 
+
+Apart from file name, we can override the default logging pattern with the property logging.pattern.file:
+``` 
+# Logging pattern for file
+logging.pattern.file= %d{yyyy-MM-dd HH:mm:ss} [%thread] %-5level %logger{36} - %msg%
+
+```
+
+Other properties related to file :   
+
+
+| Property        | What It Means           | Value If Not Set  |
+| --------------------- |:-------------| :-------------:|
+| logging.file.max-size      | total size of log archive | 10 Mb |
+| logging.file.max-history      | how many days' rotated log files to be kept      |   7 Days |
+| logging.file.total-size-cap | total size of log archives. Backups are deleted when the total size of log archives exceeds that threshold.      |   Not Specified |
+| logging.file.clean-history-on-start | force log archive cleanup on application startup      |    false |
+
+
+We can apply the same customization in a separate file which we will see in the next section. 
+
+### Isolate Logging Configuration From Application With logback Configuration
+ 
+We can isolate the log configuration from the application by specifying the configuration in logback.xml or logback-spring in XML or groovy syntax. Spring recommends using the names with spring-like logback-spring.xml or logback-spring.groovy.
+The configuration is comprised of appender element inside a root configuration tag. The pattern is specified inside an encoder element :
+
+```xml
+<configuration >
+  <include
+    resource="/org/springframework/boot/logging/logback/base.xml" />
+  <appender name="STDOUT"
+    class="ch.qos.logback.core.ConsoleAppender">
+    <encoder>
+      <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n
+      </pattern>
+    </encoder>
+  </appender>
+```
+
+Logback uses a configuration library - Joran so we will see these logs during application startup if we set a debug property in the configuration tag to true.
+
+```xml
+<configuration debug="true">
+```
+
+## Making Our Logging Useful
+We need to capture relevant information in our logs for our logging to be useful. Let us look at few of those:
+### What do we log
```

Comment:
This is good advice, but the list is rather random. I'd rather move this into a section "Why is Logging Important" in the beginning of this article. In that section, don't just list what to log like here, but instead come from a use case perspective: when do we need logging? When we're analyzing bugs or when we want to count business events and display them on a dashboard or ... then weave the items from this list in there, if it fits.